R/utils.R

```diff
@@ -803,3 +803,179 @@ doPrunePed <- function(ped, addMissingMates=FALSE, solveMultiGraph="use.all"){
+#' E-Learning Platform by Friedrich Pahlke, Second Edition.
+#' \doi{10.1002/9783527633654.ch6}.
+#'
+#' @seealso [PedigreeUtils] for other pedigree utility functions.
+#'
+#' @md
+#'
+#' @examples
+#'
+#' ##########################
+#' ##
+#' ##  Create a new FAData object
+#' ##
+#' ## Load the Minnesota Breast Cancer record and subset to the
+#' ## first families.
+#' data(minnbreast)
+#' mbsub <- minnbreast[minnbreast$famid %in% 1:20, ]
+#' mbped <- mbsub[, c("famid", "id", "fatherid", "motherid", "sex")]
+#' ## Renaming column names
+#' colnames(mbped) <- c("family", "id", "father", "mother", "sex")
+#' ## Defining the optional argument age.
+#' Age <- mbsub$endage
+#' names(Age) <- mbsub$id
+#' ## Create the object
+#' fad <- FAData(pedigree=mbped, age=Age)
+#'
+#' ## Getting all pairs of individuals with a kinship coefficient >= 0.25
+#' ## keeping all duplicates
+#' rel_pairs <- kinshipPairs(fad)
+#' head(rel_pairs)
+#' ## As we see, we have multiple times the individual 1 etc.
+#'
+#' ## For an actual correlation analysis it would be better to drop duplicates.
+#' ## Below we randomly select individual pairs if they occurr multiple times
+#' rel_pairs <- kinshipPairs(fad, duplicates = "random")
+#' head(rel_pairs)
+#'
+#' ## In addition we extract pairs of individuals that are much less related.
+#' ## For this examples we consider all individuals with a kinship
+#' ## coefficient < 0.03125 (second cousin) to be *unrelated*.
+#' unrel_pairs <- kinshipPairs(fad, duplicates = "random",
+#'     condition = function(z) z < 0.03125)
+#' head(unrel_pairs)
+#'
+#' ## For a familial resemblance analysis we can now calculate the correlation
+#' ## coefficient of a quantitative trait between pairs of related individuals
+#' ## and compare that with the correlation coefficient calculated on unrelated
+#' ## individuals. For our toy example we use the participant's age, since we
+#' ## don't have any other quantitative values available.
+#' cor_rel <- cor(age(fad)[rel_pairs[, 1]], age(fad)[rel_pairs[, 2]],
+#'     use = "pairwise.complete.obs")
+#'
+#' cor_unrel <- cor(age(fad)[unrel_pairs[, 1]], age(fad)[unrel_pairs[, 2]],
+#'     use = "pairwise.complete.obs")
+#' cor_rel
+#' cor_unrel
+#'
+#' ## We don't see a clear difference in the correlation, thus, the age (as
+#' ## expected) age has no familial component.
+kinshipPairs <- function(x, condition = function(x) x >= 0.25,
+                         duplicates = c("keep", "first", "last", "random"),
+                         id = NULL, family = NULL) {
+    duplicates <- match.arg(duplicates)
+    if (!inherits(x, "FAData"))
+        stop("'kinshipPairs' is currently only implemented for 'FAData'")
+    if (length(family)) {
+        id <- x$id[x$family %in% family]
+        if (!length(id))
+            stop("No individuals for the provided families found")
+    }
+    ks <- kinship(x)
+    if (length(id)) {
+        is_id <- rownames(ks) %in% id
+        ks <- ks[is_id, is_id]
```

Comment:
Just like above with the families, you want to check if after subsetting there is still something left to work on.

Reply:
good point. I will fix.

Reply:
Great, thanks.